### PR TITLE
Add cpu argument to Dataset API

### PR DESCRIPTION
**Pull request by rjzamora**
Addresses the first two items in #534 

This PR introduces the `cpu` argument to the `Dataset` API (and to the `DatasetEngine` backends).  It also introduces the `to_cpu` and `to_gpu` methods.  Note that none of the operators (nor `Dataset.to_parquet`) support a pandas-backed partition at the moment, so any use of `cpu=True` or `to_cpu()` will result in a user warning.  Overall, this is just the first step towards a "CPU version" of NVTabular.

### CODE REVIEW

File: nvtabular/io/dataset.py [hunk 1]
```diff
@@ -18,9 +18,13 @@
 import random
 import warnings
 
-import cudf
+try:
+    import cudf
+    import dask_cudf
+except ImportError:
+    cudf = None
+    dask_cudf = None
```
Comment (rjzamora) [inline nvtabular/io/dataset.py:26]: Not sure if we should wait for "full" cpu support before covering the case that cudf/dask_cudf is not installed...

Comment (benfred) [reply nvtabular/io/dataset.py:26]: Yeah - I think fully handling the case where there is no cudf/dask_cudf will require changes to almost every module here and is probably out of scope of the 0.4 release.  I think we can wait on this for the 'full' support in 0.5

**Comment by benfred** [comment]
rerun tests

File: nvtabular/io/csv.py [hunk 2]
```diff
@@ -36,8 +36,23 @@ def __init__(self, paths, part_size, storage_options=None, **kwargs):
...
+        if cpu:
+            if columns:
+                return dd.read_csv(self.paths, blocksize=self.part_size, **self.csv_kwargs)[columns]
+            return dd.read_csv(self.paths, blocksize=self.part_size, **self.csv_kwargs)
+        else:
+            if columns:
+                return dask_cudf.read_csv(self.paths, chunksize=self.part_size, **self.csv_kwargs)[
+                    columns
+                ]
+            return dask_cudf.read_csv(self.paths, chunksize=self.part_size, **self.csv_kwargs)
```
Comment (benfred) [inline nvtabular/io/csv.py:52]: super minor nitpick: I think it'd be cleaner to minimize the number of branches here - maybe something like this instead?

```suggestion
        if cpu:
            ddf = dd.read_csv(self.paths, blocksize=self.part_size, **self.csv_kwargs)
        else:
            ddf = dask_cudf.read_csv(self.paths, chunksize=self.part_size, **self.csv_kwargs)
        if columns:
            ddf = ddf[columns]
        return ddf
```

File: nvtabular/io/parquet.py [hunk 3]
```diff
@@ -109,18 +110,33 @@ def num_rows(self):
...
+            except ValueError:
+                return dd.read_parquet(
+                    self.paths,
+                    engine="pyarrow",
+                    columns=columns,
+                    index=None if columns is None else False,
+                    gather_statistics=False,
+                    split_row_groups=self.row_groups_per_part,
+                    storage_options=self.storage_options,
+                )
```
Comment (benfred) [inline nvtabular/io/parquet.py:139]: What's the cause of this fallback? Is this due to different versions of dask? If so - can you add a comment here explaining why this is necessary

(I'm not against having this code here - just want to make sure I understand why its there =)

**Review by benfred — APPROVED** [review]
lgtm

File: nvtabular/io/parquet.py (hunk 3 above)
Comment (rjzamora) [reply nvtabular/io/parquet.py]: Right - Older versions of Dask will not support the `"pyarrow-dataset"` engine.  This code is indeed confusing, so I tried to clean it up a bit.

**Review by benfred — APPROVED** [review]

**Pull request merged**